File: Kotlin/app/src/main/kotlin/problems/Problem125.kt
import Utils.*

// https://leetcode.com/problems/valid-palindrome/description/
fun isPalindrome(s: String): Boolean {
    if (s.length == 1) return true

    var l = 0
    var r = s.length - 1

    while (l != r && l < r) {
        val chL = s[l]
        val chR = s[r]

        if (!chL.isDigit() && !chL.isLetter()) {
            l++
            continue
        }

        if (!chR.isDigit() && !chR.isLetter()) {
            r--
            continue
        }

        if (chL.isDigit() || chL.isLetter() && chR.isDigit() || chR.isLetter()) {
            if (chL.lowercaseChar() != chR.lowercaseChar()) return false
        }

        l++
        r--
    }

    return true
}

fun test125() {
    println("125) Valid Palindrome:")
    val ts =
        mapOf(
            "A man, a plan, a canal: Panama" to true,
            "race a car" to false,
            " " to true,
        )

    for ((input, expected) in ts) {
        val res = expected == isPalindrome(input)
        println("For $input: ${colorResult(res)}")
    }
}

File: Dotnet/Problems/Problem125.cs
namespace Problems;

public sealed class Problem125
{
    // https://leetcode.com/problems/valid-palindrome/description/
    public bool IsPalindrome(string s)
    {
        if (s.Length == 1) return true;

        var l = 0;
        var r = s.Length - 1;

        while (l != r && l < r)
        {
            var chL = s[l];
            var chR = s[r];

            if (!char.IsLetterOrDigit(chL))
            {
                l++;
                continue;
            }

            if (!char.IsLetterOrDigit(chR))
            {
                r--;
                continue;
            }

            if (char.IsLetterOrDigit(chL) && char.IsLetterOrDigit(chR))
                if (char.ToLowerInvariant(chL) != char.ToLowerInvariant(chR))
                    return false;

            l++;
            r--;
        }

        return true;
    }
}
